File: googleDriveClone-kys/googleDrive/src/main/java/com/googledrive/file/FileController.java
package com.googledrive.file;

import com.googledrive.common.ApiResponse;
import com.googledrive.file.dto.CreateFolderRequest;
import com.googledrive.file.dto.FileResponse;
import com.googledrive.file.dto.MoveRequest;
import com.googledrive.file.dto.RenameRequest;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.Map;

/**
 * 파일/폴더 관련 REST API 컨트롤러.
 *
 * 모든 응답은 ApiResponse<T> 래퍼로 감싸서 반환한다.
 * 현재 로그인한 사용자 ID는 SecurityContextHolder에서 추출한다.
 */
@RestController
@RequestMapping("/api/files")
@RequiredArgsConstructor
public class FileController {

    private final FileService fileService;

    // ──────────────────────────────────────────
    // 특수 문서함 (경로 충돌 방지를 위해 {id} 매핑보다 먼저 선언)
    // ──────────────────────────────────────────

    /**
     * 최근 문서함.
     * GET /api/files/recent
     */
    @GetMapping("/recent")
    public ResponseEntity<ApiResponse<List<FileResponse>>> listRecent() {
        Long userId = getCurrentUserId();
        return ResponseEntity.ok(ApiResponse.success(fileService.listRecent(userId)));
    }

    /**
     * 별표(중요) 문서함.
     * GET /api/files/starred
     */
    @GetMapping("/starred")
    public ResponseEntity<ApiResponse<List<FileResponse>>> listStarred() {
        Long userId = getCurrentUserId();
        return ResponseEntity.ok(ApiResponse.success(fileService.listStarred(userId)));
    }

    /**
     * 공유 문서함 (미구현 — 빈 리스트 반환).
     * GET /api/files/shared
     */
    @GetMapping("/shared")
    public ResponseEntity<ApiResponse<List<FileResponse>>> listShared() {
        Long userId = getCurrentUserId();
        return ResponseEntity.ok(ApiResponse.success(fileService.listShared(userId)));
    }

    /**
     * 휴지통 목록.
     * GET /api/files/trash
     */
    @GetMapping("/trash")
    public ResponseEntity<ApiResponse<List<FileResponse>>> listTrash() {
        Long userId = getCurrentUserId();
        return ResponseEntity.ok(ApiResponse.success(fileService.listTrash(userId)));
    }

    // ──────────────────────────────────────────
    // 파일/폴더 기본 CRUD
    // ──────────────────────────────────────────

    /**
     * 파일/폴더 목록 조회.
     * GET /api/files?parentId=xxx  (parentId 없으면 루트)
     */
    @GetMapping
    public ResponseEntity<ApiResponse<List<FileResponse>>> listFiles(
            @RequestParam(required = false) Long parentId) {
        Long userId = getCurrentUserId();
        return ResponseEntity.ok(ApiResponse.success(fileService.listFiles(userId, parentId)));
    }

    /**
     * 폴더 생성.
     * POST /api/files/folder
     */
    @PostMapping("/folder")
    public ResponseEntity<ApiResponse<FileResponse>> createFolder(
            @Valid @RequestBody CreateFolderRequest request) {
        Long userId = getCurrentUserId();
        FileResponse response = fileService.createFolder(request, userId);
        return ResponseEntity.status(HttpStatus.CREATED).body(ApiResponse.success(response));
    }

    /**
     * 파일 업로드.
     * POST /api/files/upload
     * multipart/form-data: file (MultipartFile), parentId (optional)
     */
    @PostMapping("/upload")
    public ResponseEntity<ApiResponse<FileResponse>> uploadFile(
            @RequestParam("file") MultipartFile file,
            @RequestParam(value = "parentId", required = false) Long parentId) {
        Long userId = getCurrentUserId();
        FileResponse response = fileService.uploadFile(file, parentId, userId);
        return ResponseEntity.status(HttpStatus.CREATED).body(ApiResponse.success(response));
    }

    /**
     * 파일/폴더 상세 조회 (접근 기록 upsert 포함).
     * GET /api/files/{id}
     */
    @GetMapping("/{id}")
    public ResponseEntity<ApiResponse<FileResponse>> getFile(@PathVariable Long id) {
        Long userId = getCurrentUserId();
        return ResponseEntity.ok(ApiResponse.success(fileService.getFile(id, userId)));
    }

    /**
     * 이름 변경.
     * PATCH /api/files/{id}/rename
     */
    @PatchMapping("/{id}/rename")
    public ResponseEntity<ApiResponse<FileResponse>> renameFile(
            @PathVariable Long id,
            @Valid @RequestBody RenameRequest request) {
        Long userId = getCurrentUserId();
        return ResponseEntity.ok(ApiResponse.success(fileService.renameFile(id, request, userId)));
    }

    /**
     * 이동.
     * PATCH /api/files/{id}/move
     */
    @PatchMapping("/{id}/move")
    public ResponseEntity<ApiResponse<FileResponse>> moveFile(
            @PathVariable Long id,
            @RequestBody MoveRequest request) {
        Long userId = getCurrentUserId();
        return ResponseEntity.ok(ApiResponse.success(fileService.moveFile(id, request, userId)));
    }

    /**
     * 복사 (FILE만 지원).
     * POST /api/files/{id}/copy
     */
    @PostMapping("/{id}/copy")
    public ResponseEntity<ApiResponse<FileResponse>> copyFile(@PathVariable Long id) {
        Long userId = getCurrentUserId();
        FileResponse response = fileService.copyFile(id, userId);
        return ResponseEntity.status(HttpStatus.CREATED).body(ApiResponse.success(response));
    }

    /**
     * 다운로드 Presigned URL 반환.
     * GET /api/files/{id}/download
     * 응답 형식: {"downloadUrl": "https://..."}
     */
    @GetMapping("/{id}/download")
    public ResponseEntity<ApiResponse<Map<String, String>>> downloadFile(@PathVariable Long id) {
        Long userId = getCurrentUserId();
        String url = fileService.getDownloadUrl(id, userId);
        return ResponseEntity.ok(ApiResponse.success(Map.of("downloadUrl", url)));
    }

    /**
     * 미리보기 Presigned URL 반환 (이미지 등 inline 표시용).
     * GET /api/files/{id}/view
     * 응답 형식: {"viewUrl": "https://..."}
     */
    @GetMapping("/{id}/view")
    public ResponseEntity<ApiResponse<Map<String, String>>> viewFile(@PathVariable Long id) {
        Long userId = getCurrentUserId();
        String url = fileService.getViewUrl(id, userId);
        return ResponseEntity.ok(ApiResponse.success(Map.of("viewUrl", url)));
    }

    /**
     * 별표 토글.
     * PATCH /api/files/{id}/star
     */
    @PatchMapping("/{id}/star")
    public ResponseEntity<ApiResponse<FileResponse>> toggleStar(@PathVariable Long id) {
        Long userId = getCurrentUserId();
        return ResponseEntity.ok(ApiResponse.success(fileService.toggleStar(id, userId)));
    }

    /**
     * 휴지통으로 이동 (소프트 삭제).
     * DELETE /api/files/{id}
     */
    @DeleteMapping("/{id}")
    public ResponseEntity<ApiResponse<Void>> trashFile(@PathVariable Long id) {
        Long userId = getCurrentUserId();
        fileService.trashFile(id, userId);
        return ResponseEntity.ok(ApiResponse.success());
    }

    // ──────────────────────────────────────────
    // 휴지통 관리
    // ──────────────────────────────────────────

    /**
     * 휴지통에서 복원.
     * PATCH /api/files/{id}/restore
     */
    @PatchMapping("/{id}/restore")
    public ResponseEntity<ApiResponse<FileResponse>> restoreFile(@PathVariable Long id) {
        Long userId = getCurrentUserId();
        return ResponseEntity.ok(ApiResponse.success(fileService.restoreFile(id, userId)));
    }

    /**
     * 영구 삭제.
     * DELETE /api/files/{id}/permanent
     */
    @DeleteMapping("/{id}/permanent")
    public ResponseEntity<ApiResponse<Void>> permanentDelete(@PathVariable Long id) {
        Long userId = getCurrentUserId();
        fileService.permanentDelete(id, userId);
        return ResponseEntity.ok(ApiResponse.success());
    }

    /**
     * 휴지통 비우기 — 현재 사용자의 모든 trashed 항목을 영구 삭제한다.
     * DELETE /api/files/trash
     */
    @DeleteMapping("/trash")
    public ResponseEntity<ApiResponse<Void>> emptyTrash() {
        Long userId = getCurrentUserId();
        fileService.emptyTrash(userId);
        return ResponseEntity.ok(ApiResponse.success());
    }

    // ──────────────────────────────────────────
    // 내부 헬퍼
    // ──────────────────────────────────────────

    /**
     * SecurityContextHolder에서 현재 로그인한 사용자 ID를 추출한다.
     * JwtAuthenticationFilter에서 Authentication.setName(userId)로 저장했으므로
     * getName()을 Long으로 파싱해서 사용한다.
     */
    private Long getCurrentUserId() {
        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        return Long.parseLong(auth.getName());
    }
}

File: drive-jy/src/main/kotlin/demo/drive/file/controller/FileController.kt
package demo.drive.file.controller

import demo.drive.common.extension.currentUserId
import demo.drive.common.response.CommonResponse
import demo.drive.file.controller.dto.CreateFolderRequest
import demo.drive.file.controller.dto.FileResponse
import demo.drive.file.controller.dto.MoveRequest
import demo.drive.file.controller.dto.RenameRequest
import demo.drive.file.service.FileService
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.HttpHeaders
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

@RestController
@RequestMapping("/api/files")
class FileController(
    private val fileService: FileService,
) {

    @PostMapping("/upload")
    fun upload(
        @RequestParam("file") file: MultipartFile,
        @RequestParam("parentId", required = false) parentId: Long?,
    ): ResponseEntity<CommonResponse<FileResponse>> {
        val uploaded = fileService.upload(file, parentId, currentUserId())
        return CommonResponse.created(FileResponse.from(uploaded))
    }

    @PostMapping("/folders")
    fun createFolder(
        @RequestBody req: CreateFolderRequest,
    ): ResponseEntity<CommonResponse<FileResponse>> {
        val folder = fileService.createFolder(req.name, req.parentId, currentUserId())
        return CommonResponse.created(FileResponse.from(folder))
    }

    @GetMapping("/{id}")
    fun getFile(@PathVariable id: Long): ResponseEntity<CommonResponse<FileResponse>> {
        val file = fileService.getFile(id, currentUserId())
        return CommonResponse.ok(FileResponse.from(file))
    }

    @GetMapping("/{id}/download")
    fun download(@PathVariable id: Long, response: HttpServletResponse) {
        val (file, resource) = fileService.download(id, currentUserId())
        val encodedName = URLEncoder.encode(file.name, StandardCharsets.UTF_8).replace("+", "%20")
        val safeName = file.name.replace("\"", "\\\"")
        response.setHeader(HttpHeaders.CONTENT_DISPOSITION,
            "attachment; filename=\"$safeName\"; filename*=UTF-8''$encodedName")
        response.contentType = "application/octet-stream"
        response.setContentLengthLong(file.sizeBytes)
        resource.inputStream.use { it.copyTo(response.outputStream) }
    }

    @PatchMapping("/{id}/rename")
    fun rename(
        @PathVariable id: Long,
        @RequestBody req: RenameRequest,
    ): ResponseEntity<CommonResponse<FileResponse>> {
        val file = fileService.rename(id, req.name, currentUserId())
        return CommonResponse.ok(FileResponse.from(file))
    }

    @PatchMapping("/{id}/move")
    fun move(
        @PathVariable id: Long,
        @RequestBody req: MoveRequest,
    ): ResponseEntity<CommonResponse<FileResponse>> {
        val file = fileService.move(id, req.parentId, currentUserId())
        return CommonResponse.ok(FileResponse.from(file))
    }

    @PatchMapping("/{id}/star")
    fun toggleStar(@PathVariable id: Long): ResponseEntity<CommonResponse<FileResponse>> {
        val file = fileService.toggleStar(id, currentUserId())
        return CommonResponse.ok(FileResponse.from(file))
    }

    @GetMapping("/{id}/preview")
    fun preview(@PathVariable id: Long, response: HttpServletResponse) {
        val (file, resource) = fileService.download(id, currentUserId())
        val encodedName = URLEncoder.encode(file.name, StandardCharsets.UTF_8).replace("+", "%20")
        val safeName = file.name.replace("\"", "\\\"")
        val previewable = file.mimeType?.let {
            it.startsWith("image/") || it == "application/pdf" || it.startsWith("text/")
        } ?: false
        if (previewable) {
            response.contentType = file.mimeType!!
            response.setHeader(HttpHeaders.CONTENT_DISPOSITION,
                "inline; filename=\"$safeName\"; filename*=UTF-8''$encodedName")
        } else {
            response.contentType = "application/octet-stream"
            response.setHeader(HttpHeaders.CONTENT_DISPOSITION,
                "attachment; filename=\"$safeName\"; filename*=UTF-8''$encodedName")
        }
        response.setContentLengthLong(file.sizeBytes)
        resource.inputStream.use { it.copyTo(response.outputStream) }
    }

    @DeleteMapping("/{id}")
    fun moveToTrash(@PathVariable id: Long): ResponseEntity<CommonResponse<Nothing>> {
        fileService.moveToTrash(id, currentUserId())
        return CommonResponse.noContent()
    }
}
